### PR TITLE
[NestedTensor] Add support for transposed NestedTensors where ragged_idx > 1 for the layer_norm operator

### PR DESCRIPTION
**Pull request by jananisriram**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131520
* #131519
* #131518
* #131517

Add support for transposed, non-contiguous `NestedTensor`s, where `ragged_idx > 1`, for the `layer_norm` aten operator. This diff enables reducing along the jagged dimension for non-contiguous `NestedTensor`s, transposed between non-batch dimensions as well as between a ragged and a non-batch dimension. For example, users can now reduce a `NestedTensor` of shape `(B, M, *, N)` along `*` or `(B, N, M, *)` along `*`.

Write unit tests to verify the accuracy of the transposed ragged reduction implementation for `torch.nn.functional.layer_norm`.

Differential Revision: [D60015387](https://our.internmc.facebook.com/intern/diff/D60015387/)

### CODE REVIEW

File: torch/nested/_internal/ops.py [hunk 1]
```diff
@@ -1124,6 +1124,11 @@ def native_layer_norm_default(func, *args, **kwargs):
         "normalized_shape"
     ]  # denote -1 as the ragged dimension, i.e. normalize on the ragged dimension
 
+    if len(normalized_shape) < 2:
+        raise RuntimeError(
+            "layer_norm(): not supported for a normalized shape of length less than 2"
+        )
```
Comment (davidberard98) [inline torch/nested/_internal/ops.py:1130]: I think this breaks behavior that previously worked, right? `len(normalized_shape) == 1` seems to work when I test locally before this PR

File: torch/nested/_internal/ops.py [hunk 2]
```diff
@@ -1138,17 +1143,27 @@ def native_layer_norm_default(func, *args, **kwargs):
 
     if (
         -1 in normalized_shape
-    ):  # special handling for normalizing over the ragged dimension
+    ):  # special handling for normalizing over the ragged dimension, assuming that -1 represents the ragged shape
```
Comment (davidberard98) [inline torch/nested/_internal/ops.py:1146]: I don't think this handling is correct:

Say we have a (A, B, j0, C, D) tensor and we want to take the layernorm with normalized_shape (j0, C, D); this will actually take normalized_shape (B, j0, C, D).

**Review by davidberard98 — CHANGES_REQUESTED** [review]
See inline comments.

However, since this seems pretty hard to implement correctly and IDK if we concretely have a use case where we'd need this, we can probably skip this case
